### DISCUSSION TITLE
pokes triggering themselves

### DISCUSSION THREAD

**Discussion by cxrodgers** [wehr-lab/autopilot#60] [General]
Does anyone have issues with the pokes triggering themselves even when nothing pokes them? This happens some fraction of the time for me, maybe 20%. I hooked up a voltmeter to the output of the IR poke detector, and it reads a stable 3.3V whether or not a poke is spuriously detected. The voltage does goes to zero during an actual poke, as expected. I'm not sure if it's a hardware or software issue.. any guesses?

**Comment by cxrodgers** [comment]
Oh, i should say this doesn't happen at totally random times. It happens when a different port gets poked. So if I actually poke the center poke, there's like a 20% chance the left or right poke will immediately register a spurious poke, and an 80% chance it will function properly and detect no poke.

**Reply by cxrodgers** [reply]
Hm, also happens even if the GPIO inputs for the other pokes are tied directly to 3.3V, they can sometimes still pick up a spurious "poke". Starting to suspect a software issue.

**Comment by sneakers-the-rat** [comment]
interesting, no I've never seen this, how are you seeing the pokes be triggered? like some callback registered to them?

**Reply by cxrodgers** [reply]
i'm running the free_water task and also the nafc task, same phenomenon in each. so i'm judging this by the response of the task, not inspecting the callbacks. I do have the pilot running with DEBUG loglevel, so I can see the events flying by, but not adept enough to interpret all of them yet

**Reply by sneakers-the-rat** [reply]
another fascinating and totally critical bug. those register the `Task.handle_trigger` method as a callback here: https://github.com/wehr-lab/autopilot/blob/043fb6cf7c8642d4169be0823c773b358b2c4984/autopilot/hardware/gpio.py#L729 

which should only register callbacks for the particular pin the GPIO is set up for... the only thing I can think of would be something fishy happening if you have `record=True` but again that should only trigger from the single pin...

you might try adding a debug logging call inside handle_trigger to see if you get multiple calls to it when you trigger a single port??

**Reply by sneakers-the-rat** [reply]
it's odd because we also use those tasks regularly and trying to jam on kff-target ports is definitely part of my regular testing process

**Reply by cxrodgers** [reply]
lol will debug. I'm sure it's some weird interaction of some reasonable assumption and some innocuous decision I made in setting stuff up here. we'll learn a lot from whatever it turns out to be. On the plus side, I'm free water training mice here with autopilot!! Would like to upload a video but that no doubt violates some university policy. But you can imagine

**Reply by sneakers-the-rat** [reply]
happy debugging, lmk if it doesn't reveal itseld. as always, the possibility of there being some preposterously badly written hunk of code is distinctly nonzero ;p
